ocaml/typing/typecore.ml

```diff
@@ -3428,58 +3458,93 @@ let collect_apply_args env funct ignore_labels ty_fun ty_fun0 mode_fun sargs ret
                   eliminate_optional_arg ()
-                else begin
+                else
                   (* No argument was given for this parameter, we abstract over
                      it. *)
-                  may_warn funct.exp_loc
-                    (Warnings.Non_principal_labels "commuted an argument");
-                  Omitted { mode_fun; ty_arg; mode_arg; level = lv; sort_arg }
-                end
+                  omit_arg ()
         in
         loop ty_ret ty_ret0 mode_ret ((l, arg) :: rev_args) remaining_sargs
     | _ ->
         type_unknown_args ()
   in
   loop ty_fun ty_fun0 mode_fun [] sargs
 
-let type_omitted_parameters expected_mode env ty_ret mode_ret args =
-  let ty_ret, mode_ret, _, _, args =
+let type_omitted_parameters expected_mode env ty_ret mode_ret funct_mode args =
+  let ty_ret, mode_ret, open_args, closed_args, underscore_args, args =
     List.fold_left
-      (fun (ty_ret, mode_ret, open_args, closed_args, args) (lbl, arg) ->
+      (fun (ty_ret, mode_ret, open_args, closed_args, underscore_args, args) (lbl, arg) ->
          match arg with
          | Arg (exp, exp_mode, sort) ->
-             let open_args = (exp_mode, exp) :: open_args in
+             let open_args = (exp_mode, exp.exp_loc) :: open_args in
              let args = (lbl, Arg (exp, sort)) :: args in
-             (ty_ret, mode_ret, open_args, closed_args, args)
-         | Omitted { mode_fun; ty_arg; mode_arg; level; sort_arg } ->
+             (ty_ret, mode_ret, open_args, closed_args, underscore_args, args)
+         | Dummy { mode_fun; ty_arg; mode_arg; level; sort_arg } ->
+             let underscore_args = mode_arg :: underscore_args in
+             let args = (lbl, Dummy {mode_fun; ty_arg; mode_arg; level; sort_arg} ) :: args in
+             (ty_ret, mode_ret, open_args, closed_args, underscore_args, args)
+         | Omitted { mode_fun; ty_arg; mode_arg; level; sort_arg;} ->
              let arrow_desc = (lbl, mode_arg, mode_ret) in
              let ty_ret =
                newty2 ~level
                  (Tarrow (arrow_desc, ty_arg, ty_ret, commu_ok))
              in
              let new_closed_args =
                List.map
-                 (fun (marg, exp) ->
-                    submode ~loc:exp.exp_loc ~env ~reason:Other
+                 (fun (marg, loc) ->
+                    submode ~loc ~env ~reason:Other
                       marg (mode_partial_application expected_mode);
                     Value.regional_to_local_alloc marg)
                  open_args
              in
              let closed_args = new_closed_args @ closed_args in
              let open_args = [] in
-             let mode_closed_args = List.map Alloc.close_over closed_args in
-             let mode_partial_fun = Alloc.partial_apply mode_fun in
+             (* note that [mode_fun] is closed_over, not partial_apply *)
+             let all_closed = mode_fun :: closed_args @ underscore_args in
```

Comment:
Is there a test that we can write (perhaps involving unique/once, which you mentioned distinguish between `partial_apply` and `close_over`) that will help us avoid regressing in the future?

Reply:
In your example of `g`, when we look at `Omitted y`, the function applying to `y` is `f x`, and thus `mode_fun` describes the mode of `f x`. 

In your example of `f`, we will look at `Omitted x`, the function applying to `x` happens to be `f`, and thus `mode_fun` describes the mode of `f`.

Reply:
As for `close_over` vs `partial_apply`, consider a function
```
let unique_ f ~x ~y = ... 
```
then we have 
```
f ~y
```
unfolds to
```
fun ~x -> f ~x ~y 
```
And because `f` is `unique`, this function closing over it should be `once`.

But of course I'm wrong; this closing over of `f` is result of unfolding and has specific shape. In particular, it doesn't use any capability provided by the uniqueness of `f`. That is, we can think this function closes over weakened-to-`shared` version of `f` and doesn't need to be `once`. 

I will fix this.

Reply:
We discussed this and agree the new code is right. I'll suggest a comment in review in a moment.